Add Masked Scatter

CC-Yeh opened the pull request:
## Proposed changes

Supports masked scatter via indexing.

```python
self[mask] = src
```

For the CPU path: iterate with contiguous iterators over `mask`/`out` (and per-batch when `vmap_axis` == 0), copying the next `src` element whenever `mask` is true.

For the GPU path: It first computes `scatter_offsets` with `cumsum`, then use `scatter_offsets` + `mask` to do the assignment in a second kernel.

Closes #2149

Benchmarks results On Apple M4 chip.

<img width="720" height="864" alt="Apple_M4_masked_scatter_float32" src="https://github.com/user-attachments/assets/d8cc41e4-188f-4b2d-acea-474ff06210c3" />


## Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

awni commented on tests/ops_tests.cpp:
```diff
@@ -2422,6 +2422,44 @@ TEST_CASE("test scatter") {
+    auto self = zeros({4}, int32);
+    auto mask = array({true, true, false, true});
+    auto source = array({1, 2, 4});
+    auto out = masked_scatter(self, mask, source, Device::cpu);
```
Why use the cpu device here only?

awni replied:
(And same for the following tests)

awni commented on mlx/backend/metal/indexing.cpp:
```diff
@@ -641,4 +643,154 @@ void ScatterAxis::eval_gpu(const std::vector<array>& inputs, array& out) {
 }
 
+void MaskedScatter::eval_gpu(const std::vector<array>& inputs, array& out) {
+  const array& self = inputs[0];
```
This is a nit but I think it makes sense to change the name `self` to `out` (or `dst` or something similar). The term `self` is a bit overloaded from Python so it's slightly confusing.

awni commented on mlx/backend/metal/indexing.cpp:
```diff
@@ -641,4 +643,154 @@ void ScatterAxis::eval_gpu(const std::vector<array>& inputs, array& out) {
+  set_u32(as_u32(total), bind_idx++);
+  set_u32(is_batched_run ? as_u32(inner) : 0u, bind_idx++);
+  set_u32(src_has_batches ? as_u32(src_block) : 0u, bind_idx++);
+  set_u32(is_batched_run ? as_u32(src_capacity) : 0u, bind_idx++);
```
Nit: I think the `set_u32` is unnecessary. Cleaner to just call `compute_encoder.set_bytes(...) ` directly.

awni commented on mlx/primitives.cpp:
```diff
@@ -4348,6 +4348,107 @@ bool ScatterAxis::is_equivalent(const Primitive& other) const {
+          self.shape(),
+          tangents[1].dtype(),
+          std::make_shared<MaskedScatter>(stream()),
+          {out, mask_b, tangents[1]});
```
I'm confused by this. It looks like you are applying the same scatter twice? Am I missing something?

awni commented on mlx/primitives.cpp:
```diff
@@ -4348,6 +4348,107 @@ bool ScatterAxis::is_equivalent(const Primitive& other) const {
+
+std::pair<std::vector<array>, std::vector<int>> MaskedScatter::vmap(
+    const std::vector<array>& inputs,
+    const std::vector<int>& axes) {
```
The vmap of a masked scatter should ideally be a masked scatter with appropriatley transformed inputs. E.g you may need to broadcast / repeat the mask. We should aim to remove the vmap axis from the primitive as I don't think it's necessary. Instead try to do the vmap using a regular masked scatter.

CC-Yeh replied:
Removed `vmap_axis_`, and added default batch for general case.
This really simplified the implementation, thanks!

awni commented:
The results are looking quite impressive! Sorry for the delay in getting it reviewed. I left a few comments. Could you please take a look and address when you have time?

angeloskath commented on mlx/primitives.cpp:
```diff
@@ -1317,15 +1317,15 @@ Shape Convolution::conv_out_shape(
-      msg << "[conv] Padding sizes must be non-negative." << " Got padding "
-          << pads_lo << " | " << pads_hi << ".";
+      msg << "[conv] Padding sizes must be non-negative."
+          << " Got padding " << pads_lo << " | " << pads_hi << ".";
```
This is a total nitpick but since it's in the diff would you mind fixing it more correctly by joining the string literals?

CC-Yeh commented on mlx/backend/metal/indexing.cpp:
```diff
@@ -641,4 +643,154 @@ void ScatterAxis::eval_gpu(const std::vector<array>& inputs, array& out) {
+  if (src_batched && (src_size % batch != 0)) {
+    throw std::invalid_argument(
+        "[MaskedScatter::eval_gpu] src.size() must be divisible by batch when src is batched.");
+  }
```
That make senses.
Moved some to ops and the rest are gone by introducing default batching.

CC-Yeh commented on mlx/primitives.cpp:
```diff
@@ -4348,6 +4348,107 @@ bool ScatterAxis::is_equivalent(const Primitive& other) const {
+        const array idx_vec = reshape(idx_src, {n}, s);
+        const array cotan_updates = reshape(cotan_src, {n, 1}, s);
+        gsrc_flat = scatter_add(gsrc_flat, idx_vec, cotan_updates, 0, s);
+      }
```
I think a `gather` solution would be slower and more complex because, without `nonzero` or `argwhere` APIs, we'd probably have to perform the same complex `cumsum` and `scatter_add` before `gather` to get compact indices array.

If you'd anything in mind, I'm happy to implement it)

CC-Yeh commented:
Hey @awni 

Thanks again for the review, learned a lot from it!

I did some improvements. Can you take a look again?

angeloskath commented on mlx/backend/metal/kernels/indexing/masked_scatter.h:
```diff
@@ -0,0 +1,47 @@
+  }
+
+  const uint src_index = scatter_offsets[elem_to_loc<uint>(
+      idx, offsets_shapes, offsets_strides, offsets_ndim)];
```
Unless I am mistaken, since we flattened the mask, the offsets are an output of scan (which is contiguous but we can also check and copy if needed) and the output is a contiguous array we can probably remove all `elem_to_loc` calls except the one for `src`.

angeloskath reviewed (changes requested):
Very nice indeed! Thanks for the thorough work!

I left a spattering of comments but a more significant concern is that it doesn't implement the same API as numpy and I think we should. Boolean indices in numpy do not really broadcast and omitted dimensions are taken whole.

Basically
```python
x = np.arange(1000).reshape(10, 10, 10)
x[np.random.randn(10, 10) > 0] = 0    # valid
x[np.random.randn(1, 10, 10) > 0] = 0 # invalid
x[np.random.randn(10, 10, 1) > 0] = 0 # invalid
```

This also allows for a more efficient implementation for the cases where the axes being copied as is are the large ones. I do not think that you need to optimize this but I think we should implement the API at least.

The first and simplest approach would be to expand and broadcast the mask in the op then the primitive should work on `dst` and `mask` that have the same size and shape (which pretty much is already a requirement).

CC-Yeh commented:
@angeloskath 

Thanks for the insightful suggestions! Learned a lot from it 😃 

I addressed most comments, looking forward to your thoughts, thanks.

angeloskath commented:
Ok finally fixed another misconception in vmap (added a double vmap test that shows what wasn't working). It was there before but also in the vmap I wrote.

Basically the `MaskedScatter` primitive is batched. The 0-th dim is the batch. The `masked_scatter` op is not batched. So in the primitive we can't use the op but we have to use the primitive and also we have to deal with the previous batch.

CC-Yeh commented:
@angeloskath  Nice catch!

Fixed another test, hope it'll pass now 🙏
